lib/Dialect/Calyx/Transforms/CompileControl.cpp

```diff
@@ -0,0 +1,192 @@
+//===- CompileControl.cpp - Compile Control Pass ----------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the Compile Control pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/Calyx/CalyxOps.h"
+#include "circt/Dialect/Calyx/CalyxPasses.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OperationSupport.h"
+#include "mlir/IR/PatternMatch.h"
+
+using namespace circt;
+using namespace calyx;
+using namespace mlir;
+
+/// A helper function to create constants in the HW dialect.
+static hw::ConstantOp createConstant(OpBuilder &builder, Location loc,
+                                     size_t bitWidth, size_t value) {
+  return builder.create<hw::ConstantOp>(
+      loc, APInt(bitWidth, value, /*unsigned=*/true));
+}
+
+/// Given some number of states, returns the necessary bit width
+/// TODO(Calyx): Probably a better built-in operation?
+static size_t getNecessaryBitWidth(size_t numStates) {
+  APInt apNumStates(64, numStates);
+  size_t log2 = apNumStates.ceilLogBase2();
+  return log2 > 1 ? log2 : 1;
+}
+
+/// Creates a RegisterOp, with input and output port bit widths defined by
+/// `width`.
+static RegisterOp createRegister(OpBuilder &builder, ComponentOp &component,
+                                 size_t width, StringRef name) {
+  IRRewriter::InsertionGuard guard(builder);
+  builder.setInsertionPointToStart(component.getBody());
+  auto *context = builder.getContext();
+  return builder.create<RegisterOp>(component->getLoc(),
+                                    StringAttr::get(context, name), width);
+}
+
+/// Generates a latency-insensitive FSM to realize a sequential operation. This
+/// is done by initializing GroupGoOp values for the enabled groups in the
+/// SeqOp, and then creating a new Seq GroupOp with the given FSM. Each step in
+/// the FSM is guarded by the done operation of the group currently being
+/// executed. After the group is complete, the FSM is incremented. This SeqOp is
+/// then replaced in the control with an Enable statement referring to the new
+/// Seq GroupOp.
+static void visitSeqOp(SeqOp &seq, ComponentOp &component) {
+  auto wires = component.getWiresOp();
+  Block *wiresBody = wires.getBody();
+
+  auto &seqOps = seq.getBody()->getOperations();
+  if (!llvm::all_of(seqOps, [](auto &&op) { return isa<EnableOp>(op); }))
+    seq.emitOpError("should only contain EnableOps in this pass.");
+
```

Comment:
Should this return early at this point? I'm not sure if the rest of the pass makes sense if this invariant is not met.

Reply:
Good point. Fixed.